android/app/src/main/java/fi/fmi/mobileweather/LargeWidgetProvider.java

```diff
@@ -0,0 +1,173 @@
+package fi.fmi.mobileweather;
+
+import static android.view.View.GONE;
+import static android.view.View.VISIBLE;
+
+import android.app.PendingIntent;
+import android.content.Intent;
+import android.graphics.Bitmap;
+import android.graphics.BitmapFactory;
+import android.graphics.Color;
+import android.util.Log;
+import android.widget.RemoteViews;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.text.DateFormat;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Date;
+import java.util.Iterator;
+
+public class LargeWidgetProvider extends BaseWidgetProvider {
+    @Override
+    protected int getLayoutResourceId() {
+        return R.layout.large_widget_layout;
+    }
+
+    @Override
+    protected void onPostExecute(JSONObject json, JSONArray json2, RemoteViews main, SharedPreferencesHelper pref) {
+
+        Intent intent = new Intent(context, MainActivity.class);
+        PendingIntent pendingIntent = PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_IMMUTABLE);
+
+        // Get settings
+        String background = pref.getString("background", "transparent");
+        Log.d("Download json", "Background: " + background);
+
+        main = new RemoteViews(context.getPackageName(), getLayoutResourceId());
+
+        // Show normal view
+        main.setInt(R.id.normalLayout, "setVisibility", VISIBLE);
+        // Hide error view
+        main.setInt(R.id.errorLayout, "setVisibility", GONE);
+
+        main.setOnClickPendingIntent(R.id.mainLinearLayout, pendingIntent);
+
+        json = useNewOrStoredJsonObject(json, pref);
+        if (json == null) return;
+
+        if (background.equals("dark")) {
+            setColors(main,
+                    Color.parseColor("#191B22"),
+                    Color.rgb(255, 255, 255));
+        }
+        else if (background.equals("light")) {
+            setColors(main,
+                    Color.rgb(255, 255, 255),
+                    Color.rgb(48, 49, 147));
+        } else {
+            setColors(main,
+                    Color.TRANSPARENT,
+                    Color.rgb(48, 49, 147));
+        }
+        Log.d("Download json", "Forecast json: " + json);
+        Log.d("Download json", "Crisis json: " + json2);
+
+
+        try {
+            // Get the keys of the JSONObject
+            Iterator<String> keys = json.keys();
+
+            // Retrieve the first key
+            if (!keys.hasNext()) {
+                return;
+            }
+            String firstKey = keys.next();
+            Log.d("Download json", "First key (geoid): " + firstKey);
+
+            // Extract the JSONArray associated with the first key
+            JSONArray data = json.getJSONArray(firstKey);
+
+            // get the first 5 JsonObject from the JSONArray
```

Comment:
This should be fixed to take "first 5 of items that has epochtime in future". Because this doesn't work for cached data and can use timesteps in the past.

Reply:
Done.
Corresponding fix done also for small widgets.